File: utility_tools/flb_old/flb_io.c
/* flb_io.c 

26 aug 98 REK Changes for V5.2 and MS C

*/


#  include "stdlib.h"
#  include "stdio.h"
#  include "string.h"
#  include "filelb.h"
#  include "debug.h"

	void read_vms_flb_details (FILE *flb_ptr, FLB_DETAILS *details);

void write_flb_details (flb_ptr, details)
FILE *flb_ptr;
FLB_DETAILS *details;
{

  QUARTET string_len, vnum;
  INT_T i, num_names;

  vnum = 2;
  fwrite (&vnum, sizeof(QUARTET), 1, flb_ptr);
  fwrite (&(details->nchan), sizeof(QUARTET), 1, flb_ptr);
  fwrite (&(details->length), sizeof(QUARTET), 1, flb_ptr);
  fwrite (&(details->format), sizeof(QUARTET), 1, flb_ptr);
  string_len = strlen (details->domain_name);
  fwrite (&string_len, sizeof(QUARTET), 1, flb_ptr);
  fwrite ((details->domain_name), sizeof(char), string_len, flb_ptr);
  fwrite (&(details->increment), sizeof(float), 1, flb_ptr); 
  fwrite (&(details->domain_start), sizeof(float), 1, flb_ptr);
  string_len = strlen (details->comment);
  fwrite (&string_len, sizeof(QUARTET), 1, flb_ptr);
  fwrite ((details->comment), sizeof(char), string_len, flb_ptr);
  if (details->nchan > FLB_CHAN_MAX) {
    num_names = 1;
    }
  else {
    num_names = details->nchan;
  }
  for (i=0; i < num_names; i++){
    string_len = strlen (details->name[i]);
    fwrite (&string_len, sizeof(QUARTET), 1, flb_ptr);
    fwrite ((details->name[i]), sizeof(char),string_len, flb_ptr);
  }
  fwrite ((details->min), sizeof (double), num_names, flb_ptr);
  fwrite ((details->max), sizeof (double), num_names, flb_ptr);
  return;
}



INT_T read_flb_details ( flb_ptr, details )


     FILE  *flb_ptr;
     FLB_DETAILS *details;
{

  INT_T err_flag = -1;
  QUARTET string_len;
  INT_T i, num_names;
  if(fread (&(details->version_num), sizeof(QUARTET), 1, flb_ptr) != 1) {
  DEBUG (
	DEBUG_LOC("read_flb_details");
	 printf ("eof\n");
 )
    return err_flag;
     };

  DEBUG(
	printf("\n");
  	DEBUG_LOC("read_flb_details");
        printf("details->version_num = %d\n\n" COMMA details->version_num); 
  )

  if (details->version_num == 1) {
    read_vms_flb_details ( flb_ptr, details );
  }
  else if (details->version_num ==2){
    fread (&(details->nchan), sizeof(QUARTET), 1, flb_ptr);
    fread (&(details->length), sizeof(QUARTET), 1, flb_ptr);
    fread (&(details->format), sizeof(QUARTET), 1, flb_ptr);
    fread (&string_len, sizeof(QUARTET), 1, flb_ptr);
    if (string_len > FLB_NAME_MAX) {
      printf ("Domain name too long: %s\n", details->domain_name);
      return -1;
    }

    fread ((details->domain_name), sizeof(char), string_len, flb_ptr);
    details->domain_name[string_len] = '\0';
    fread (&(details->increment), sizeof(float), 1, flb_ptr); 
    fread (&(details->domain_start), sizeof(float), 1, flb_ptr);
    fread (&string_len, sizeof(QUARTET), 1, flb_ptr);
    if (string_len > FLB_COMMENT_MAX) {
      printf ("Comment too long: %s\n", details->comment);
      return -1;
    }
    fread ((details->comment), sizeof(char), string_len, flb_ptr);
    details->comment[string_len]= '\0';
    if (details->nchan > FLB_CHAN_MAX) {
      num_names = 1;
      }
    else {
      num_names = details->nchan;
	}
    for (i=0; i < num_names; i++){
      fread (&string_len, sizeof(QUARTET), 1, flb_ptr);
      if (string_len > FLB_NAME_MAX) {
	printf ("Channel name too long: %s\n", details->name[i]);
        return -1;
      }
      fread ((details->name[i]), sizeof(char), string_len, flb_ptr);
    details->name[i][string_len] = '\0';
    }
    fread ((details->min), sizeof (double), num_names, flb_ptr);
    fread ((details->max), sizeof (double), num_names, flb_ptr);
  }
  else {
    printf ("Bad FLB version number\n");
    return -1;
  }
  return 1;
}




void read_vms_flb_details ( flb_ptr, details )

     FILE  *flb_ptr;
     FLB_DETAILS *details;
{
  INT_T i, count;

  fread (&(details->nchan), sizeof(QUARTET), 1, flb_ptr);
  fread (&(details->length), sizeof(QUARTET), 1, flb_ptr);
  if (details->nchan > 1 ) {
    fseek ( flb_ptr, 4*(details->nchan -1), SEEK_CUR);
  }
  fread (&(details->format), sizeof(QUARTET), 1, flb_ptr);
  if (details->nchan > 1 ) {
    fseek ( flb_ptr, 4*(details->nchan -1), SEEK_CUR);
  }
  fread ((details->domain_name), 8, 1, flb_ptr);
  details->domain_name[8]= '\0';
  if (details->nchan > 1 ) {
    fseek ( flb_ptr, 8*(details->nchan -1), SEEK_CUR);
  }
  fread (&(details->increment), sizeof(float), 1, flb_ptr); 
  if (details->nchan > 1 ) {
    fseek ( flb_ptr, 4*(details->nchan -1), SEEK_CUR);
  }

  fread (&(details->domain_start), sizeof(float), 1, flb_ptr);
   if (details->nchan > 1 ) {
    fseek ( flb_ptr, 4*(details->nchan -1), SEEK_CUR);
  }

  fseek ( flb_ptr, 4* details->nchan, SEEK_CUR);
  for (i=0; i< details->nchan; i++){
    fread (details->name[i], 1,16, flb_ptr);
    details->name[i][16] = '\0';
  }
  fread (details->comment, sizeof(char), 80, flb_ptr);
  details->comment[79]='\0';
  count = 88 + (details->nchan)*44;
  fseek ( flb_ptr, 512-count, SEEK_CUR);
  return;
}

File: utility_tools/flbio.m
function D = flbio( fid, option, D )
% FLBIO - perform io operations on flb files
%
%   option:
%       'read_details - read details of next case
%       'read_data' - read data for next case
%       'read_case'
%       'read_index'
%       'read_all'
%       'write_case'
%       'new_real'
%
% Usage
%
% Example
%
% See also
%
%% AUTHOR: Robert Kearney
%% $DATE: 02-Feb-2006 08:48:32 $
%% $Revision: 1.4 $
%% DEVELOPED: 7.1.0.246 (R14) Service Pack 3
%% FILENAME: flbio.m

% DL made changes to lines:75,87,88

switch lower(option),
    case 'read_details'
        D=flbrdet (fid);
    case 'read_case'
        D=flbrdet (fid);
        if isempty(D),
            if feof(fid),
                error('End of file');
            else
                error('Error reading file');
            end
        end
        D=flbrdata(fid, D);
    case 'read_index'
        caseNum=0;
        D=struct([]);
        while 1,
            tempD =flbrdet(fid);
            if isempty(tempD),
                break
            else
                nByte = tempD.nChan * tempD.chanLen *tempD.nReal * tempD.chanFormat;
                status = fseek(fid, nByte, 'cof');
                caseNum=caseNum+1;
                if caseNum==1,
                    D=tempD;
                else
                    D(caseNum)=tempD;
                end
            end
        end
    case 'read_all'
        caseNum=0;
        while 1,
            tempD =flbrdet(fid);
            if isempty(tempD),
                break
            else
                tempD=flbrdata(fid, tempD);
                caseNum=caseNum+1;
                D(caseNum)=tempD;
            end
        end
        
    case 'skip_case'
        D=flbSkipCase(fid);
    case 'write_case'
        dataSize=size(D.Data);
        % DL changed nDim -> D.nDim
        D.nDim=length(dataSize);
        
        % Number of realizations
        if D.nDim==3,
            D.nReal=dataSize(3);
        else
            D.nReal=1;
        end
        D.nChan=dataSize(2);
        D.chanLen=dataSize(1);
        %DL changed chanMax/Min -> D.chanMax/Min
        D.chanMax=max(D.Data);
        D.chanMin=min(D.Data);
        s=flbwdet(fid,D);
        count = fwrite(fid, D.Data, 'float');
        D=1;
    case 'new_real'
        caseIdx=flbio(fid,'read_index');
        nCase=length(caseIdx);
        frewind(fid);
        for i=1:nCase-1,
            s=flbio(fid,'skip_case');
        end
        casePos=ftell(fid);
        oldDetails= flbio(fid,'read_details');
        % Check that details match
        [nSamp, nChan, nReal]=size(D.Data);
        if nSamp ~= oldDetails.chanLen,
            error ('Realization has a different sample length');
        end
        if nChan ~= oldDetails.nChan,
            error ('Realization has a differnet number channels');
        end
        %% Update Details
        Details=oldDetails;
        Details.nDim=3;
        Details.nReal=Details.nReal+ nReal;
        fseek(fid, casePos, 'bof');
        s=flbwdet(fid, Details);
        fseek(fid,0,'eof');
        count = fwrite(fid, D.Data, 'float');
        D=1;
        
    otherwise
        error (['flbio option not defined: ' option ]);
end

function D = flbrdet( fid)
% FLBRDET ...
%  Read next set of details from a flb file
%   ...
% Usage
%
% Example
%
% See also
% read details from a flb file
flb_version = fread(fid,1,'int32');
if isempty(flb_version),
    D=[];
    return
end

if (flb_version < 2),
    disp ('Header error')
    num_chan=-2;
    return
end
if flb_version== 2,
    D.nDim=2;
    D.nReal=1;
    D.nChan=fread(fid,1,'int32')
    D.chanLen=fread(fid,1,'int32')
    D.chanFormat=fread(fid,1,'int32')
    strLen=fread(fid,1,'int32')
    D.domainName=char(fread(fid,strLen,'char'))'
    D.domainIncrement=fread(fid,1,'real*4')
    D.domainStart=fread(fid,1,'real*4')
    commentLen=fread(fid,1,'int32')
    D.comment=char(fread(fid,commentLen,'char'))'
    for iChan=1:D.nChan,
        nameLen=fread(fid,1,'int32')
        D.chanName{iChan}=char(fread(fid,nameLen,'char'))'
    end
    D.chanMin=fread(fid,D.nChan,'real*8')
    D.chanMax=fread(fid,D.nChan,'real*8')
else
    D.nDim=fread(fid,1,'int32');
    D.nReal=fread(fid,1,'int32');
    D.nChan=fread(fid,1,'int32');
    D.chanLen=fread(fid,1,'int32');
    D.chanFormat=fread(fid,1,'int32');
    stringLen=fread(fid,1,'int32');
    D.domainName=char(fread(fid,stringLen,'char'))';
    D.domainIncr = fread(fid,1,'single');
    D.domainStart = fread(fid,1,'single');
    stringLen=fread(fid,1,'int32');
    D.comment=char(fread(fid,stringLen,'char'))';
    
    for iChan=1:D.nChan,
        stringLen=fread(fid,1,'int32');
        D.chanName{iChan}=char(fread(fid,stringLen,'char'))';
    end
    D.chanMin =fread(fid,D.nChan,'float64');
    D.chanMax =fread (fid,D.nChan,'float64');
end
return

%%

function D= flbrdata (fid, D)
% Read next data set
nSamp = D.nChan * D.chanLen *D.nReal;

switch D.chanFormat
    case 2
        formatSpec='short';
    case 4
        formatSpec='float';
end
tempC= fread(fid, nSamp, formatSpec);
C = reshape (tempC, D.chanLen, D.nChan, D.nReal);
D.Data=C;
return

%% Skip case

function status = flbSkipCase (fid)
D=flbrdet(fid);
nByte = D.nChan * D.chanLen *D.nReal * D.chanFormat;
status = fseek(fid, nByte, 'cof');
return

%% Write details

function status = flbwdet( fid, D)
% FLBRDET ...
%  Write a case to current file location
%  D is a structre with fields:
%       D.comment
%       D.chanName
%       D.domainIncr
%       D.domainName
%       D.domainStat
%       D.Datahelp
%   ...
% Usage
%
% Example
%
% See also
% read details from a flb file
flbVersion =  4;
count= fwrite(fid,flbVersion,'int32');
%% Number of dimensions

count = fwrite(fid,D.nDim,'int32');
count=fwrite(fid,D.nReal,'int32');
% number of channels
count=fwrite(fid,D.nChan,'int32');
% Channel Length
count=fwrite(fid,D.chanLen,'int32');
% Data Format
count=fwrite(fid,4,'int32');
% DomainName
stringLen=length(D.domainName);
count=fwrite(fid, stringLen, 'int32');
count=fwrite(fid, D.domainName,'char');
% Domain Increment
count=fwrite (fid,D.domainIncr,'single');
count=fwrite (fid,D.domainStart,'single');
% Comment
stringLen=length(D.comment);
count=fwrite(fid,stringLen,'int32');
count=fwrite(fid,D.comment,'char');
% Channel Names
for iChan=1:D.nChan,
    stringLen=length(D.chanName{iChan});
    count=fwrite(fid,stringLen,'int32');
    count=fwrite(fid, D.chanName{iChan},'char');
end

count=fwrite(fid,D.chanMin,'float64');
count=fwrite(fid,D.chanMax,'float64');


%% data



status=1;
return













